Type embeds

surma opened the pull request:
Alright, I got it working. There’s still lots to do, but it’s working, and that’s pretty damn cool.

## So what’s in here

At the time of opening PR, it’s doing the following this:

- Adding a `transform.js` that works with `asc`’s `--transform` flag. This transformer extracts the type information of every imported and exported function and puts a type descriptor into a custom section in the resulting Wasm module. It also auto-injects `as-bind.ts`, so the user doesn’t need to do that anymore.
- Delete a lot of code, as we don’t need to guess and cache types anymore. 
- Use `browserslist` to make sure we publish modern JS.
- Updating to most recent Rollup and switching from Closure to Terser. The latter I am only doing because I used to use optional chaining and Closure didn’t like that.

## What still needs to be done

- [x] Expose the `jsToAscConverters` and `ascToJsConverters` (or similar) so that users can handle conversions of their types themselves. (Note to self: Refactor this to a `converters` array with `{matcher, jsToAsc(), ascToJs()}`.)
- [x] Remove unnecessary abstractions which could make the final code even smaller.
- [x] Fix the Markdown parser demo.
- [x] Tests? Like, I think the transformer might need some tests that it catches the right types, especially in more complicated setups.
- [x] Make this work with the browser SDK
- [x] Update README
- [x] If browser SDK is proving difficult, restore the previous `as-bind` and ship both

## Future work 

- [ ] Allow memory _views_, not just copies. Maybe we can tie this to the type system (as in a function returning a `Uint8Array` vs `Uint8ArrayView`, which is just a type alias).
- [ ] In the same vein: Automatically create live class instance bindings

## Demo

```ts
declare function alert(s: string): void;
// This now works!! A string as the return type for an imported function!
declare function toUpperCase(s: string): string;

export function fill(v: u8): Uint8Array {
  const arr = new Uint8Array(32);
  arr.fill(v);
  return arr;
}

export function sum(arr: Float64Array): f64 {
  let sum: f64 = 0;
  for(let i = 0; i < arr.length; i++) {
    sum += arr[i];
  }
  return sum;
}

export function myalert(s: string): void {
  alert(toUpperCase(s) + "!!!!!");
}
```

```
$ npx asc --transform ./transform.js --exportRuntime -b demo.wasm demo.ts
```

```html
<!doctype html>
<script type="module">
  import * as AsBind from "/dist/as-bind.esm.js";

  async function main() {
    const instance = await AsBind.instantiate(fetch("/demo.wasm"), {
      demo: {
        toUpperCase: s => s.toUpperCase(),
        alert
      }
    });
    console.log(instance.exports.fill(101));
    console.log(instance.exports.sum(new Float32Array([1,2,3,4])));
    instance.exports.myalert("topkek");
  }
  main();
</script>
```

torch2424 commented on lib/asbind-instance/asbind-instance.js:
```diff
@@ -26,7 +24,32 @@ const traverseObjectAndRunCallbackForFunctions = (
+}
+
+function extractTypeDescriptor(module) {
+  const sections = WebAssembly.Module.customSections(module, SECTION_NAME);
```
Woah, I didn't know you could grab custom sections from JS :joy:

surma replied:
Neither did I!! All credit to @dcodeio here

torch2424 commented on lib/asbind-instance/bind-function.js:
```diff
@@ -1,276 +1,180 @@
-        argumentsWithReplacedRefs.push(arg);
-      }
-    });
+const ascToJsConverters = new Map([
```
OooOoOoOoO this is rad haha! :joy:

surma replied:
Glad you like it :D In [another branch](https://github.com/torch2424/as-bind/blob/type-embed-continue/test/tests/custom-type/test.js#L3-L19) I have exposed that map so you can add handlers for your own types (like if you want to polyfill `externref` for example).

torch2424 reviewed (approved):
This all looks great so far to me! :smile: 

So, in the AS Discord I asked:

> Does anyone happen to know if the browser SDK for the AS compiler supports transforms?

pointing to: https://github.com/AssemblyScript/assemblyscript/tree/master/lib/sdk

and I know there is a AsBind user who depends on the Browser SDK. But, looking at this new compiler transform version, I don't think it would be too difficult to support both? :thinking: 

So, I'd like to suggest, instead of ripping out all of the type caching / guessing code. Perhaps we can offer a compile time asbind (default), but still provide a runtime asbind (Which can have it's own README, that we just link to in the main README). And to keep the same codesize benefits of this asbind, let's make them two seperate entrypoints? :smile: 

What do you think? :smile:

surma commented:
> So, I'd like to suggest, instead of ripping out all of the type caching / guessing code. Perhaps we can offer a compile time asbind (default), but still provide a runtime asbind (Which can have it's own README, that we just link to in the main README). And to keep the same codesize benefits of this asbind, let's make them two seperate entrypoints? 😄

That not only makes sense but is actually really reasonable. I’ll whip that up once I land the Test PR.

Re running transforms in the browser SDK: I’ll give that a try! I have no idea about the browser SDK, so that’s gonna be interesting :D

trusktr commented:
I see the demo of the string-returning function. How does the type information relate? Can you provide a demo of why that's needed?

surma commented:
> I see the demo of the string-returning function. How does the type information relate? Can you provide a demo of why that's needed?

Hm I am not quite sure I follow. But the reason I highlight the imported function returning a string is that those weren’t handled by as-bind previously. The type information makes sure that as-bind knows what type is expected and can do the right conversion from JS to Wasm linear memory and vice versa. Does that answer your question?

dcodeIO commented on lib/asbind-instance/type-converters.js:
```diff
@@ -91,7 +91,10 @@ export const converters = new Map([
   // it seems unlikely for that to be a problem.
-  [/^(i|u|f)(8|16|32|64)$/, { ascToJs: nop, jsToAsc: nop }],
+  [
+    /^(i|u|f)(8|16|32|64)|[ui]size|bool|externref$/,
```
Someone might declare their own `type f8 = something`. Would that be a problem?

surma replied:
No, because I am looking at the resolved type. 

(I actually think you sometimes might want to declare a type alias and associate different handling with it, but I’ll solve that another day)

surma commented:
Alright! This is ready to go imo!

- Update README
- Updated WebSDK demo to use the new option added by @dcodeIO  (makes it a lot simpler, thanks!)
- Fixed CI

torch2424 reviewed (approved):
Yooo! Thank you so much for all of this work @surma !!! :smile: :tada: 

And thanks for the assist @dcodeIO ! :smile: 

So! I'm gonna pin this tab in my Chrome, and do a merge and release tomorrow for this! :smile:

torch2424 commented:
Deployed in `0.7.0` :smile: :tada: Thank you @surma !!!! :smile:
